neuralop/layers/coda_blocks.py

```diff
@@ -0,0 +1,398 @@
+        self.temperature = temperature
+        self.num_dims = len(n_modes)
+
+        if Normalizer is None:
+            Normalizer = partial(nn.InstanceNorm2d, affine=True) if self.num_dims == 2 else partial(
+                nn.InstanceNorm3d, affine=True)
+
+        if spectral_convolution is None:
+            spectral_convolution = SpectralConv
+
+        # K,Q,V operator with or without non_liniarity
+        if nonlinear_attention:
+            kqv_activation = non_linearity
+        else:
+            kqv_activation = torch.nn.Identity()
+
+        self.permutation_eq = permutation_eq
+
+        self.codimension_size = codimension_size
+        self.mixer_token_codimension = token_codimension
+
+        if per_channel_attention:
+            # for per channel attention, forcing the values of token dims
+            self.token_codimension = 1
+            self.head_codimension = 1
+
+        # this scale used for downsampling Q,K functions
+        if scale is None:
+            scale = 2 if per_channel_attention else 1
+            scale = min(self.n_heads, scale)
+
+        mixer_modes = [i // scale for i in n_modes]
+
+        if decomposition_kwargs is None:
+            decomposition_kwargs = {}
+        common_args = dict(
+            use_channel_mlp=use_channel_mlp,
+            preactivation=preactivation,
+            channel_mlp_skip=channel_mlp_skip,
+            mlp_dropout=0,
+            incremental_n_modes=incremental_n_modes,
+            rank=rank,
+            fft_norm=fft_norm,
+            channel_mlp_expansion=channel_mlp_expansion,
+            fixed_rank_modes=fixed_rank_modes,
+            implementation=implementation,
+            separable=separable,
+            factorization=factorization,
+            decomposition_kwargs=decomposition_kwargs,
+            joint_factorization=joint_factorization,
+        )
+
+        kqv_args = dict(
+            in_channels=self.token_codimension,
+            out_channels=self.n_heads * self.head_codimension,
+            n_modes=mixer_modes,
+            # args below are shared with Projection block
+            non_linearity=kqv_activation,
+            fno_skip='linear',
+            norm=None,
+            apply_skip=True,
+            n_layers=1,
+        )
+        self.K = FNOBlocks(
+            output_scaling_factor=1 / scale,
+            conv_module=spectral_convolution,
+            **kqv_args,
+            **common_args,
+        )
+        self.Q = FNOBlocks(
+            output_scaling_factor=1 / scale,
+            conv_module=spectral_convolution,
+            **kqv_args,
+            **common_args,
+        )
+        self.V = FNOBlocks(
+            output_scaling_factor=1,
+            conv_module=spectral_convolution,
+            **kqv_args,
+            **common_args,
+        )
+
+        if self.n_heads * self.head_codimension != self.token_codimension:
+            self.proj = FNOBlocks(
```

Comment:
This name collides with the projection in FNO (that is an MLP on channels). 

Perhaps we can make it more explicitly that it is an FNOBLock not a simple projection layer.